multi/stm32l4-multi/libmulti/libdma.c

```diff
@@ -11,205 +11,452 @@
 
-	_configureChannel(dma, channel, dir, priority, paddr, msize, psize, minc, pinc, reqmap);
+	if (flags == 0) {
+		return -1;
+	}
 
-	return 0;
+	*(base + ifcr) = (0xF << (channel * 4));
+
+	switch (dma_transfers[dma][channel].type) {
+		case dma_transferOnce:
+			libdma_unprepareTransfer(dma_transfers[dma][channel].once.channelBase);
+
+			*dma_transfers[dma][channel].once.doneFlag = 1;
+
+			dma_transfers[dma][channel].type = dma_transferNull;
+			break;
+		case dma_transferInf:
+			if ((flags & DMA_HTIE_FLAG) != 0) {
+				dma_transfers[dma][channel].inf.fn(dma_transfers[dma][channel].inf.arg, dma_ht);
+			}
+			if ((flags & DMA_TCIE_FLAG) != 0) {
+				dma_transfers[dma][channel].inf.fn(dma_transfers[dma][channel].inf.arg, dma_tc);
+			}
+			break;
+		case dma_transferNull:
+		default:
+			/* Shouldn't happen */
+			break;
+	}
+
+	return 1;
 }
 
 
-static void _prepare_transfer(volatile unsigned int *channel_base, void *maddr, size_t len, int tcie)
+static void libdma_configureChannel(int dma, int channel, int dir, int priority, void *paddr, int msize, int psize, int minc, int pinc, unsigned char reqmap, handle_t *cond)
 {
-	*(channel_base + cmar) = (unsigned int) maddr;
-	*(channel_base + cndtr) = len;
-	dataBarier();
+	unsigned int tmp, irqnum = libdma_irqnum(dma, channel);
+	volatile unsigned int *channelBase = dma_common[dma].base + 2 + (5 * channel);
+	handle_t interruptCond = (cond == NULL) ? dma_common[dma].cond : *cond;
 
-	if (tcie)
-		/* Enable Transfer Complete interrupt, enable channel */
-		*(channel_base + ccr) |= (1 << 1) | 0x1;
-	else
-		/* Enable channel */
-		*(channel_base + ccr) |= 0x1;
+	interrupt(irqnum, libdma_irqHandler, (void *)dma, interruptCond, NULL);
 
-	dataBarier();
+	*(channelBase + ccr) = ((priority & 0x3) << 12) | ((msize & 0x3) << 10) | ((psize & 0x3) << 8) |
+		((minc & 0x1) << 7) | ((pinc & 0x1) << 6) | ((dir & 0x1) << 4);
+	*(channelBase + cpar) = (unsigned int)paddr;
+	tmp = *(dma_common[dma].base + cselr) & ~(0xF << channel * 4);
+	*(dma_common[dma].base + cselr) = tmp | ((unsigned int)reqmap << channel * 4);
 }
 
 
-static void _unprepare_transfer(volatile unsigned int *channel_base)
+int libdma_configurePeripheral(const struct libdma_per *per, int dir, int priority, void *paddr, int msize, int psize, int minc, int pinc, handle_t *cond)
 {
-	dataBarier();
-	/* Disable Transfer Complete interrupt, disable channel */
-	*(channel_base + ccr) &= ~((1 << 1) | 0x1);
-	dataBarier();
+	int dma = per->dma, channel = per->channel[dir];
+	char reqmap = per->reqmap;
+
+	libdma_configureChannel(dma, channel, dir, priority, paddr, msize, psize, minc, pinc, reqmap, cond);
+
+	return 0;
 }
 
 
-static int _has_channel_finished(const void *maddr, volatile unsigned int *channel_base)
+static int libdma_hasChannelFinished(volatile unsigned int *channelBase)
 {
-	return (maddr != NULL && *(channel_base + cndtr) > 0);
+	return (*(channelBase + cndtr) > 0) ? 1 : 0;
```

Comment:
This might not work when no rx buffer has been provided (`rxMAddr`) in function `libdma_transferHelper`. Also it kind of looks like this function has a reversed logic - when channel has finished it returns 0 and when channel has not finished it returns 1. Might be a good idea to change it for better readability